<a href="https://colab.research.google.com/github/DanielRaiicHu/telecom_x_2/blob/main/telecom_x_2.ipynb?forceRefresh=true">
  <img src="https://drive.google.com/uc?export=view&id=191Woxwut0iaIL2hCcCIeXRSB6oba_88H" alt="Abrir en Colab" width="120"/>
</a>

# **Proyecto: Telecom X - Parte 2**

## **Preparación de los Datos**

### **Extracción del Archivo Tratado**

Se importan y tratan los datos generados en el desafío anterior

In [82]:
import pandas as pd

In [83]:
url = 'https://raw.githubusercontent.com/DanielRaiicHu/telecom_x_2/main/datos_tratados.csv'
datos = pd.read_csv(url)
datos.head()

ID_Cliente Evasion     Genero AdultoMayor TienePareja Dependientes  \
0  0002-ORFBO      No   Femenino          No          Sí           Sí   
1  0003-MKNFE      No  Masculino          No          No           No   
2  0004-TLHLJ      Sí  Masculino          No          No           No   
3  0011-IGKFF      Sí  Masculino          Sí          Sí           No   
4  0013-EXCHZ      Sí   Femenino          Sí          Sí           No   

   MesesAntiguedad ServicioTelefonico LineasMultiples ServicioInternet  ...  \
0                9                 Sí              No              DSL  ...   
1                9                 Sí              Sí              DSL  ...   
2                4                 Sí              No     Fibra óptica  ...   
3               13                 Sí              No     Fibra óptica  ...   
4                3                 Sí              No     Fibra óptica  ...   

  ProteccionDispositivo SoporteTecnico StreamingTV StreamingPeliculas  \
0                    No             Sí          Sí                 No   
1                    No             No          No                 Sí   
2                    Sí             No          No                 No   
3                    Sí             No          Sí                 Sí   
4                    No             Sí          Sí                 No   

  TipoContrato FacturaElectronica          MetodoPago CuentasDiarias  \
0       Un año                 Sí   Cheque por correo           2.19   
1      Mensual                 No   Cheque por correo           2.00   
2      Mensual                 Sí  Cheque electrónico           2.46   
3      Mensual                 Sí  Cheque electrónico           3.27   
4      Mensual                 Sí   Cheque por correo           2.80   

  CargoMensual  CargoTotal  
0         65.6      593.30  
1         59.9      542.40  
2         73.9      280.85  
3         98.0     1237.85  
4         83.9      267.40  

[5 rows x 22 columns]

In [84]:
datos.shape

(7043, 22)

In [85]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ID_Cliente             7043 non-null   object 
 1   Evasion                7043 non-null   object 
 2   Genero                 7043 non-null   object 
 3   AdultoMayor            7043 non-null   object 
 4   TienePareja            7043 non-null   object 
 5   Dependientes           7043 non-null   object 
 6   MesesAntiguedad        7043 non-null   int64  
 7   ServicioTelefonico     7043 non-null   object 
 8   LineasMultiples        7043 non-null   object 
 9   ServicioInternet       7043 non-null   object 
 10  SeguridadEnLinea       7043 non-null   object 
 11  RespaldoEnLinea        7043 non-null   object 
 12  ProteccionDispositivo  7043 non-null   object 
 13  SoporteTecnico         7043 non-null   object 
 14  StreamingTV            7043 non-null   object 
 15  Stre

In [86]:
datos.describe()

MesesAntiguedad  CuentasDiarias  CargoMensual   CargoTotal
count      7043.000000     7043.000000   7043.000000  7043.000000
mean         32.371149        2.158675     64.761692  2279.734304
std          24.559481        1.003088     30.090047  2266.794470
min           0.000000        0.610000     18.250000     0.000000
25%           9.000000        1.180000     35.500000   398.550000
50%          29.000000        2.340000     70.350000  1394.550000
75%          55.000000        2.990000     89.850000  3786.600000
max          72.000000        3.960000    118.750000  8684.800000

In [87]:
datos.describe(include = 'O')

ID_Cliente Evasion     Genero AdultoMayor TienePareja Dependientes  \
count         7043    7043       7043        7043        7043         7043   
unique        7043       2          2           2           2            2   
top     9995-HOTOH      No  Masculino          No          No           No   
freq             1    5174       3555        5901        3641         4933   

       ServicioTelefonico LineasMultiples ServicioInternet SeguridadEnLinea  \
count                7043            7043             7043             7043   
unique                  2               3                3                3   
top                    Sí              No     Fibra óptica               No   
freq                 6361            3390             3096             3498   

       RespaldoEnLinea ProteccionDispositivo SoporteTecnico StreamingTV  \
count             7043                  7043           7043        7043   
unique               3                     3              3           3   
top                 No                    No             No          No   
freq              3088                  3095           3473        2810   

       StreamingPeliculas TipoContrato FacturaElectronica          MetodoPago  
count                7043         7043               7043                7043  
unique                  3            3                  2                   4  
top                    No      Mensual                 Sí  Cheque electrónico  
freq                 2785         3875               4171                2365

In [88]:
# Revisión de datos únicos para cada columna
for col in datos.columns:
    if datos[col].dtype == 'object':
        print(f"Valores únicos en '{col}':")
        print(datos[col].unique())
        print("—" * 40)

Valores únicos en 'ID_Cliente':
['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
————————————————————————————————————————
Valores únicos en 'Evasion':
['No' 'Sí']
————————————————————————————————————————
Valores únicos en 'Genero':
['Femenino' 'Masculino']
————————————————————————————————————————
Valores únicos en 'AdultoMayor':
['No' 'Sí']
————————————————————————————————————————
Valores únicos en 'TienePareja':
['Sí' 'No']
————————————————————————————————————————
Valores únicos en 'Dependientes':
['Sí' 'No']
————————————————————————————————————————
Valores únicos en 'ServicioTelefonico':
['Sí' 'No']
————————————————————————————————————————
Valores únicos en 'LineasMultiples':
['No' 'Sí' 'Sin servicio telefónico']
————————————————————————————————————————
Valores únicos en 'ServicioInternet':
['DSL' 'Fibra óptica' 'No']
————————————————————————————————————————
Valores únicos en 'SeguridadEnLinea':
['No' 'Sí' 'Sin servicio de internet']
——————————————

### **Eliminación de columnas irrelevantes**

In [91]:
datos = datos.drop('ID_Cliente', axis=1)
datos.head()

Evasion     Genero AdultoMayor TienePareja Dependientes  MesesAntiguedad  \
0      No   Femenino          No          Sí           Sí                9   
1      No  Masculino          No          No           No                9   
2      Sí  Masculino          No          No           No                4   
3      Sí  Masculino          Sí          Sí           No               13   
4      Sí   Femenino          Sí          Sí           No                3   

  ServicioTelefonico LineasMultiples ServicioInternet SeguridadEnLinea  ...  \
0                 Sí              No              DSL               No  ...   
1                 Sí              Sí              DSL               No  ...   
2                 Sí              No     Fibra óptica               No  ...   
3                 Sí              No     Fibra óptica               No  ...   
4                 Sí              No     Fibra óptica               No  ...   

  ProteccionDispositivo SoporteTecnico StreamingTV StreamingPeliculas  \
0                    No             Sí          Sí                 No   
1                    No             No          No                 Sí   
2                    Sí             No          No                 No   
3                    Sí             No          Sí                 Sí   
4                    No             Sí          Sí                 No   

  TipoContrato FacturaElectronica          MetodoPago CuentasDiarias  \
0       Un año                 Sí   Cheque por correo           2.19   
1      Mensual                 No   Cheque por correo           2.00   
2      Mensual                 Sí  Cheque electrónico           2.46   
3      Mensual                 Sí  Cheque electrónico           3.27   
4      Mensual                 Sí   Cheque por correo           2.80   

   CargoMensual  CargoTotal  
0          65.6      593.30  
1          59.9      542.40  
2          73.9      280.85  
3          98.0     1237.85  
4          83.9      267.40  

[5 rows x 21 columns]

### **Tratamiento de los datos binarios**

In [92]:
import warnings
# Ignoramos mensajes de advertencia para mantener la salida limpia
warnings.filterwarnings('ignore')

In [93]:
datos_codificados = datos.copy()

# Columnas con Sí/No
columnas_sin_no = ['Evasion','AdultoMayor','TienePareja','Dependientes','ServicioTelefonico','FacturaElectronica']

# Transformación a las variables de las variables con Sí/No. 'Sí' = 1 y 'No' = 0
datos_codificados[columnas_sin_no] = datos_codificados[columnas_sin_no].replace({'Sí':1, 'No':0})

# Transformación de la variable Genero
datos_codificados.Genero = datos_codificados.Genero.replace({'Femenino':0,'Masculino':1})

In [94]:
datos_codificados

Evasion  Genero  AdultoMayor  TienePareja  Dependientes  \
0           0       0            0            1             1   
1           0       1            0            0             0   
2           1       1            0            0             0   
3           1       1            1            1             0   
4           1       0            1            1             0   
...       ...     ...          ...          ...           ...   
7038        0       0            0            0             0   
7039        1       1            0            1             0   
7040        0       1            0            0             0   
7041        0       1            0            1             1   
7042        0       1            0            1             1   

      MesesAntiguedad  ServicioTelefonico          LineasMultiples  \
0                   9                   1                       No   
1                   9                   1                       Sí   
2                   4                   1                       No   
3                  13                   1                       No   
4                   3                   1                       No   
...               ...                 ...                      ...   
7038               13                   1                       No   
7039               22                   1                       Sí   
7040                2                   1                       No   
7041               67                   1                       No   
7042               63                   0  Sin servicio telefónico   

     ServicioInternet SeguridadEnLinea  ... ProteccionDispositivo  \
0                 DSL               No  ...                    No   
1                 DSL               No  ...                    No   
2        Fibra óptica               No  ...                    Sí   
3        Fibra óptica               No  ...                    Sí   
4        Fibra óptica               No  ...                    No   
...               ...              ...  ...                   ...   
7038              DSL               Sí  ...                    No   
7039     Fibra óptica               No  ...                    No   
7040              DSL               No  ...                    No   
7041              DSL               Sí  ...                    Sí   
7042              DSL               Sí  ...                    Sí   

     SoporteTecnico StreamingTV StreamingPeliculas TipoContrato  \
0                Sí          Sí                 No       Un año   
1                No          No                 Sí      Mensual   
2                No          No                 No      Mensual   
3                No          Sí                 Sí      Mensual   
4                Sí          Sí                 No      Mensual   
...             ...         ...                ...          ...   
7038             Sí          No                 No       Un año   
7039             No          No                 Sí      Mensual   
7040             No          No                 No      Mensual   
7041             Sí          No                 Sí     Dos años   
7042             No          Sí                 Sí     Dos años   

     FacturaElectronica          MetodoPago CuentasDiarias  CargoMensual  \
0                     1   Cheque por correo           2.19         65.60   
1                     0   Cheque por correo           2.00         59.90   
2                     1  Cheque electrónico           2.46         73.90   
3                     1  Cheque electrónico           3.27         98.00   
4                     1   Cheque por correo           2.80         83.90   
...                 ...                 ...            ...           ...   
7038                  0   Cheque por correo           1.84         55.15   
7039                  1  Cheque electrónico           2.84         85.10   
7040                  1   Cheque por correo           1.68        

In [95]:
datos_codificados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Evasion                7043 non-null   int64  
 1   Genero                 7043 non-null   int64  
 2   AdultoMayor            7043 non-null   int64  
 3   TienePareja            7043 non-null   int64  
 4   Dependientes           7043 non-null   int64  
 5   MesesAntiguedad        7043 non-null   int64  
 6   ServicioTelefonico     7043 non-null   int64  
 7   LineasMultiples        7043 non-null   object 
 8   ServicioInternet       7043 non-null   object 
 9   SeguridadEnLinea       7043 non-null   object 
 10  RespaldoEnLinea        7043 non-null   object 
 11  ProteccionDispositivo  7043 non-null   object 
 12  SoporteTecnico         7043 non-null   object 
 13  StreamingTV            7043 non-null   object 
 14  StreamingPeliculas     7043 non-null   object 
 15  Tipo

In [96]:
# Las variables que siguen siendo tipo objecty las almacena en categoricas_restantes
categoricas_restantes =  datos_codificados.select_dtypes(include='object').columns.tolist()

In [97]:
datos_codificados = pd.get_dummies(data = datos_codificados, columns=(categoricas_restantes), drop_first=True, dtype=int)

In [98]:
# Para mantener el orden original de mi dataframe
datos_codificados = datos_codificados.reindex(
    columns=sum(
        [
            ([c] if c not in categoricas_restantes else
             [col for col in datos_codificados.columns if col.startswith(f"{c}_")])
            for c in datos.columns
        ],[]
               )
            )

In [99]:
datos_codificados.head()

Evasion  Genero  AdultoMayor  TienePareja  Dependientes  MesesAntiguedad  \
0        0       0            0            1             1                9   
1        0       1            0            0             0                9   
2        1       1            0            0             0                4   
3        1       1            1            1             0               13   
4        1       0            1            1             0                3   

   ServicioTelefonico  LineasMultiples_Sin servicio telefónico  \
0                   1                                        0   
1                   1                                        0   
2                   1                                        0   
3                   1                                        0   
4                   1                                        0   

   LineasMultiples_Sí  ServicioInternet_Fibra óptica  ...  \
0                   0                              0  ...   
1                   1                              0  ...   
2                   0                              1  ...   
3                   0                              1  ...   
4                   0                              1  ...   

   StreamingPeliculas_Sí  TipoContrato_Mensual  TipoContrato_Un año  \
0                      0                     0                    1   
1                      1                     1                    0   
2                      0                     1                    0   
3                      1                     1                    0   
4                      0                     1                    0   

   FacturaElectronica  MetodoPago_Cheque por correo  \
0                   1                             1   
1                   0                             1   
2                   1                             0   
3                   1                             0   
4                   1                             1   

   MetodoPago_Tarjeta de crédito (automático)  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

   MetodoPago_Transferencia bancaria (automática)  CuentasDiarias  \
0                                               0            2.19   
1                                               0            2.00   
2                                               0            2.46   
3                                               0            3.27   
4                                               0            2.80   

   CargoMensual  CargoTotal  
0          65.6      593.30  
1          59.9      542.40  
2          73.9      280.85  
3          98.0     1237.85  
4          83.9      267.40  

[5 rows x 32 columns]

In [100]:
datos_codificados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 32 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Evasion                                         7043 non-null   int64  
 1   Genero                                          7043 non-null   int64  
 2   AdultoMayor                                     7043 non-null   int64  
 3   TienePareja                                     7043 non-null   int64  
 4   Dependientes                                    7043 non-null   int64  
 5   MesesAntiguedad                                 7043 non-null   int64  
 6   ServicioTelefonico                              7043 non-null   int64  
 7   LineasMultiples_Sin servicio telefónico         7043 non-null   int64  
 8   LineasMultiples_Sí                              7043 non-null   int64  
 9   ServicioInternet_Fibra óptica            

### **Verificación de la proporción**

Visualización gráfica de clientes que evaden (Sí) vs clientes que permanecen (No)

In [89]:
import plotly.express as px

In [151]:
cantidad_evasion = datos_codificados.Evasion.value_counts()
proporcion_evasion = datos_codificados['Evasion'].value_counts(normalize=True).sort_index()

print(f'De un total de {datos_codificados["Evasion"].size} clientes.')
print(f'Los clientes que evaden son: {cantidad_evasion[1].sum()}, equivalentes al {proporcion_evasion[1]:.2%}')
print(f'Los clientes que permanecen son: {cantidad_evasion[0].sum()}, equivalentes al {proporcion_evasion[0]:.2%}')

De un total de 7043
Los clientes que evaden son: 1869, equivalentes al 26.54%
Los clientes que permanecen son: 5174, equivalentes al 73.46%


In [90]:
px.histogram(datos, x='Evasion', text_auto=True, color='Evasion')

---
>* **Se observa una diferencia significativa entre la cantidad de clientes que permanecen y aquellos que abandonan el servicio. Aunque la mayoría de los clientes se mantiene, el porcentaje de evasión alcanza un 26.54%, lo cual representa una proporción bastante considerable.**

---

### **Separación de Variables Explicativas y Respuesta**

In [103]:
from sklearn.model_selection import train_test_split

# Variables Explicativas
X = datos_codificados.drop('Evasion', axis=1)

# Variable Respuesta
y = datos_codificados['Evasion']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [105]:
X.head()

Genero  AdultoMayor  TienePareja  Dependientes  MesesAntiguedad  \
0       0            0            1             1                9   
1       1            0            0             0                9   
2       1            0            0             0                4   
3       1            1            1             0               13   
4       0            1            1             0                3   

   ServicioTelefonico  LineasMultiples_Sin servicio telefónico  \
0                   1                                        0   
1                   1                                        0   
2                   1                                        0   
3                   1                                        0   
4                   1                                        0   

   LineasMultiples_Sí  ServicioInternet_Fibra óptica  ServicioInternet_No  \
0                   0                              0                    0   
1                   1                              0                    0   
2                   0                              1                    0   
3                   0                              1                    0   
4                   0                              1                    0   

   ...  StreamingPeliculas_Sí  TipoContrato_Mensual  TipoContrato_Un año  \
0  ...                      0                     0                    1   
1  ...                      1                     1                    0   
2  ...                      0                     1                    0   
3  ...                      1                     1                    0   
4  ...                      0                     1                    0   

   FacturaElectronica  MetodoPago_Cheque por correo  \
0                   1                             1   
1                   0                             1   
2                   1                             0   
3                   1                             0   
4                   1                             1   

   MetodoPago_Tarjeta de crédito (automático)  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

   MetodoPago_Transferencia bancaria (automática)  CuentasDiarias  \
0                                               0            2.19   
1                                               0            2.00   
2                                               0            2.46   
3                                               0            3.27   
4                                               0            2.80   

   CargoMensual  CargoTotal  
0          65.6      593.30  
1          59.9      542.40  
2          73.9      280.85  
3          98.0     1237.85  
4          83.9      267.40  

[5 rows x 31 columns]

In [106]:
y.head()

0    0
1    0
2    1
3    1
4    1
Name: Evasion, dtype: int64